File: uri/strings/html/1667.cpp
#include <iostream>
#include <string>
using namespace std;

int main() {
    string token;
    int pos = 0;
    
    while (cin >> token) {
        if (token == "<br>") {
            cout << '\n';
            pos = 0;
        } else if (token == "<hr>") {
            if (pos > 0) {
                cout << '\n';
                pos = 0;
            }
            cout << string(80, '-') << '\n';
        } else {
            int len = token.length();
            if (pos == 0) {
                cout << token;
                pos = len;
            } else if (pos + 1 + len <= 80) {
                cout << " " << token;
                pos += 1 + len;
            } else {
                cout << '\n' << token;
                pos = len;
            }
        }
    }
    
    if (pos > 0) {
        cout << '\n';
    }
    
    return 0;
}

File: uri/strings/html/1667.rs
use std::io::{self, Read, Write, BufWriter};

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    
    let mut input = String::new();
    stdin.lock().read_to_string(&mut input).unwrap();
    
    let mut pos = 0;
    
    for token in input.split_whitespace() {
        if token == "<br>" {
            if pos > 0 {
                writeln!(out).unwrap();
                pos = 0;
            }
        } else if token == "<hr>" {
            if pos > 0 {
                writeln!(out).unwrap();
                pos = 0;
            }
            writeln!(out, "{}", "-".repeat(80)).unwrap();
        } else {
            let len = token.len();
            if pos == 0 {
                write!(out, "{}", token).unwrap();
                pos = len;
            } else if pos + 1 + len <= 80 {
                write!(out, " {}", token).unwrap();
                pos += 1 + len;
            } else {
                writeln!(out).unwrap();
                write!(out, "{}", token).unwrap();
                pos = len;
            }
        }
    }
    
    if pos > 0 {
        writeln!(out).unwrap();
    }
}
